r-session-04.qmd

```diff
@@ -0,0 +1,800 @@
+    # to a selection of columns
+    across(
+      # The cols argument takes a selection of columns to apply
+      # a function to. Here, we want to apply the as.numeric()
+      # function to all columns, so we use the function
+      # everything() to select all columns.
+      .cols = everything(),
+      .fns = as.numeric
+    )
+  ) %>%
+  # Convert the dataframe from wide to long format, so we have a
+  # column for the time, a column for the state, and a column
+  # for the proportion of the population in that state at that
+  # time
+  pivot_longer(
+    # Don't pivot the time column
+    cols = -time,
+    names_to = "state",
+    values_to = "proportion"
+  ) %>%
+  # Update the state column to be a factor, so the plot will
+  # show the states in the correct order
+  mutate(state = factor(state, levels = c("S", "I", "R")))
+
+SIRcolors <- c(S = "#1f77b4", I = "#ff7f0e", R = "#FF3851")
+
+ggplot(sir_sol_df, aes(x = time, y = proportion, color = state)) +
+  geom_line(linewidth = 1.5) +
+  scale_color_manual(values = SIRcolors) +
+  labs(
+    x = "Time",
+    y = "Fraction",
+    color = "State"
+  ) +
+  theme(legend.position = "top")
+```
+
+Now, perhaps we used the tools from R-session 3 to estimate $R_0$ and we think a reasonable $95\%$ confidence interval for $R_0$ is $(1.7,1.9)$. 
+If we're quite certain the duration of infection is 6 days, then that means our corresponding confidence interval on the transmission rate is $(.283,.317)$.
+An entirely reasonble way to represent this uncertainty in the estimate of the transmission rate is to generate many random draws from within the confidence interval for the transmission rate and examine the resulting epidemic curves.
+For this we can modify the code above with a loop that does this $M$ times. 
+
+```{r}
+#| echo: true
+#| message: false
+#| out-width: 100%
+#| column: body
+library(tidyverse)
+library(deSolve)
+
+theme_set(theme_minimal())
+
+M = 100 # the number of times we want to do this. 
+        # This is a mater of choice and computational capacity (this model is small and quick, but that won't always be the case)
+
+#since we're going to do this M times we need a place to store the results
+
+for(m in 1:M){
+  sir_params <- c(transmission = runif(1,.283,.317), duration = 6) # each time take a random draw of the transmission rate from the confidence interval
+  sir_init_states <- c(S = 0.99, I = 0.01, R = 0)
+  sim_times <- seq(0, 200, by = 0.1)
+
+  sir_sol <- deSolve::ode(
+    y = sir_init_states,
+    times = sim_times,
+    func = sir_model,
+    parms = sir_params
+  )
+  sir_sol = data.frame(sir_sol)
+  sir_sol$iteration = m   # mark this as the mth iteration of the loop
+  #if this is the first iteration, create a place to store the output
+  #if it's the sencond or higher, append the output to the storage
+  if(m==1){sir_sol_storage <- sir_sol}
+  if(m>1){sir_sol_storage <- rbind(sir_sol_storage,sir_sol)}
+}
```

Comment:
This could also be written with `purrr::map()` with `purrr::list_rbind(list, names_to = "id")` (https://purrr.tidyverse.org/reference/list_c.html) to create a dataframe directly without needing the if statement.